system/src/system_power_manager.cpp

```diff
@@ -552,52 +510,87 @@ void PowerManager::handleStateChange(battery_state_t from, battery_state_t to, b
-    if (m - possibleFaultTimestamp_ > DEFAULT_FAULT_WINDOW) {
-      possibleFaultTimestamp_ = m;
-      possibleFaultCounter_ = 0;
-      faultSecondaryCounter_ = 0;
-    } else {
-      possibleFaultCounter_++;
-      if (possibleFaultCounter_ >= DEFAULT_FAULT_COUNT_THRESHOLD &&
-         (faultSuppressed_ == 0 || (m - faultSuppressed_ >= DEFAULT_FAULT_SUPPRESSION_PERIOD))) {
-        if (faultSecondaryCounter_ > 0) {
-          faultSecondaryCounter_ = 0;
-          return BATTERY_STATE_DISCONNECTED;
+void PowerManager::deduceBatteryStateLoop() {
+  if ((config_.flags & HAL_POWER_CHARGE_STATE_DISABLE)) {
+    // Charging is disabled
+    if (millis() - batMonitorTimeStamp_ > BATTERY_STATE_CHECK_PERIOD) {
+      batMonitorTimeStamp_ = millis();
+      PMIC power(true);
+      FuelGauge fuel(true);
+      // When charging is disabled, the VCELL can reliably refect the voltage on VBAT
+      float vCell = fuel.getVCell();
+      LOG_DEBUG(TRACE, "VCell: %dmV", (uint16_t)(vCell * 1000));
+      if (g_batteryState != BATTERY_STATE_DISCONNECTED) {
+        uint8_t status = power.getSystemStatus();
+        if (vCell < BATTERY_CONNECTED_VCELL_THRESHOLD && !(status & 0x08)/* MUST NOT in DPM mode, in case of sudden sink in suplement mode*/) {
+          // If battery falls below 2.0v, it's probably detached
+          battery_state_t batteryState = BATTERY_STATE_DISCONNECTED;
+          handleStateChange(g_batteryState, batteryState);
+        }
+      } else {
+        if (vCell > BATTERY_CONNECTED_VCELL_THRESHOLD) {
+          vcellDebounceCount_++;
+          if (vcellDebounceCount_ >= BATTERY_VCELL_DEBOUNCE_COUNT) {
+            vcellDebounceCount_ = 0;
+            battery_state_t batteryState = BATTERY_STATE_NOT_CHARGING;
+            handleStateChange(g_batteryState, batteryState);
+          }
         } else {
-          PMIC power;
-          power.setRechargeThreshold(PMIC_FAULT_RECHARGE_THRESHOLD);
-          possibleFaultCounter_ = 0;
-          faultSecondaryCounter_ = 1;
-          possibleFaultTimestamp_ = millis();
+          vcellDebounceCount_ = 0;
         }
       }
     }
-  }
-  return to;
-}
-
-void PowerManager::handlePossibleFaultLoop() {
-  if (faultSecondaryCounter_ == 1 && (millis() - possibleFaultTimestamp_ > DEFAULT_FAULT_WINDOW)) {
-      PMIC power;
-      power.setRechargeThreshold(PMIC_NORMAL_RECHARGE_THRESHOLD);
-      faultSecondaryCounter_ = 0;
-      faultSuppressed_ = millis();
+  } else {
+    // Charging is enabled
+    battery_state_t batteryState;
+    uint8_t status;
+    PMIC power(true);
+    FuelGauge fuel(true);
+    if (chargedTimeStamp_ != 0 && millis() - chargedTimeStamp_ >= BATTERY_CHARED_POSTPONE_PERIOD) {
+      chargedTimeStamp_ = 0;
+      status = power.getSystemStatus();
+      // Battery is charged, then the battery voltage should above VSYS_MIN (3.5v)
+      // The VSYS_STAT bit should be cleared, otherwise, the battery is disconnected
+      if (status & 0x01) {
```

Comment:
We should check that VCELL is above for example 4.0v as well, because when charging is enabled and the battery is disabled the VBAT is floating and may goes above VSYS_MIN.